llvm/lib/Support/InitLLVM.cpp

```diff
@@ -43,6 +43,10 @@ InitLLVM::InitLLVM(int &Argc, const char **&Argv,
   assert(!Initialized && "InitLLVM was already initialized!");
   Initialized = true;
 #endif
+#ifdef _WIN32
+  // Avoid searching the directory from which the application is loaded.
```

Comment:
That's not what `SetDllDirectoryA("")` does, from <https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setdlldirectorya#parameters>:
"If this parameter is an empty string (""), the call removes the current directory from the default DLL search order."

So, this is removing the current directory (not the directory where the application is loaded) from the search order: that's a good thing.

Except that Windows already prefers system and the application directory over the current directory by default: <https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-unpackaged-apps>

If removing the current directory from the search path is the desired behavior, then I'd recommend using `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_DEFAULT_DIRS)` instead, since it's much easier to understand.

Reply:
The intention is to remove the folder from which the application is loaded, not the current directory, as per the comment; thank you for catching this!

Reply:
Loading dlls from the directory where the application is loaded is not a security vulnerability: the attacker hasn't gained anything.

If the attacker has the ability to write to the directory where the application is loaded then they don't have to play games with dropping binaries with names that collide with the system - they can modify the application's binary instead.

Furthermore, changing the behavior of `LoadLibrary` doesn't help with the system libraries that are loaded statically, for that you need the `/DEPENDENTLOADFLAG` linker flag.

Reply:
For reference: <https://devblogs.microsoft.com/oldnewthing/20121031-00/?p=6203>
"Remember, the directory is the application bundle. If you drop your application into a random directory, you’ve just added everything in that directory to your bundle. And if you don’t secure your application directory, you’re allowing anybody to add components to your bundle."

Reply:
And official MSRC guidance <https://msrc.microsoft.com/blog/2018/04/triaging-a-dll-planting-vulnerability/>:
"A DLL planting issue that falls into this category, Application Directory DLL planting, is treated as Defense-in-Depth issue that will be considered for updates in future versions only. We resolve any MSRC case that fall in this category as vNext consideration, mainly due to the amount of social engineering involved in the attack and the by design nature of the bug. A victim would have to be tricked into placing the malicious DLL (malware) where it can be triggered AND perform a non-recommended action (like running an installer in the same directory as the malware). A non-installed application has no reference point for “known good directory/binaries”, unless it creates the directory itself. Ideally, the installer should create a temporary directory with a randomized name (to prevent further DLL planting), extract its binaries to it and use them to install the application. While an attacker can make use of a drive-by download to place the malware on the victim’s system, such as into the “Downloads” folder, the essence of the attack is social engineering."

Reply:
> Loading dlls from the directory where the application is loaded is not a security vulnerability: the attacker hasn't gained anything.
> 
> If the attacker has the ability to write to the directory where the application is loaded then they don't have to play games with dropping binaries with names that collide with the system - they can modify the application's binary instead.

Although that is true when we are dealing with unsigned binaries, it does not when dealing with signed ones. After modifying the binary directly, the attacker won't be able to reauthenticate it. The renamed, fake dll remains a vulnerability because it does not have to be signed in order to be loaded, so we have to ensure the search path does not find these system libraries where a user can place them.

Reply:
> Although that is true when we are dealing with unsigned binaries, it does not when dealing with signed ones. After modifying the binary directly, the attacker won't be able to reauthenticate it. The renamed, fake dll remains a vulnerability because it does not have to be signed in order to be loaded, so we have to ensure the search path does not find these system libraries where a user can place them.

Unless you are using cert-pinning or allow-listing, then the signature doesn't matter: the attacker can replace the application with a correctly signed malicious application.

Reply:
It may be the case that one has cert-pinning or allow-listing so as to be safe. And if the search path is restricted, then one can be safer.

Reply:
> If removing the current directory from the search path is the desired behavior, then I'd recommend using `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_DEFAULT_DIRS)` instead, since it's much easier to understand.

Reworked the commit to instead use LoadLibraryExA(..., LOAD_LIBRARY_SEARCH_SYSTEM32). It seems LOAD_LIBRARY_SEARCH_DEFAULT_DIRS will search user directories, among other flags listed, which we should avoid since falling back to finding dbghelp.dll or dbgcore.dll here is still dangerous. There doesn't seem to be a flag to omit just the load directory.

Reply:
Oh, and thanks a lot for the information. All of these comments were very helpful.